### ISSUE TITLE
EF Core Global Query Filter for SaveChanges (Update, Delete)

### ISSUE THREAD

**Issue by r-aghaei** [dotnet/efcore#24122]
Is there any way to apply a query filter to `WHERE` clause of generated `UPDATE` and `DELETE` statement using EF Core when you call `SaveChanges`? (Quite similar to [Global query filters](https://docs.microsoft.com/en-us/ef/core/querying/filters?WT.mc_id=DT-MVP-5003235) which work just for `SELECT` scenarios.) 

I'm going to use such feature for a multi-tenancy scenario in a small application and  I'm able to handle this requirement in other layers, like API, DAL, BLL or by overriding `SaveChanges` and checking if the deleted or modified entity belongs to the right tenant; The main reason that I prefer to not do it in BLL or SaveChanges, is because of performance. A WHERE clause will be evaluated at database server side, it's secure and also high performance, quite similar to global query filters. 

For example, for a Delete scenario, an obvious delete scenario could be something like this:

    var entity = db.BlogPosts.Find(id);
    if(entity!=null)
        db.Entry(entity).State = EntityState.Deleted;
    db.SaveChanges();

Which is basically a `SELECT` and then a `DELETE`.

However, I prefer to delete like this, which is just a `DELETE` statement:

    db.Entry(new BlogPost() { Id = id }).State = EntityState.Deleted;
    db.SaveChanges();

I've already asked a question in stackoverflow: [EF Core Global Query Filter for SaveChanges (Update, Delete)](https://stackoverflow.com/q/66125638/3110834) and answered my own question by using [Command Interceptors](https://docs.microsoft.com/en-us/ef/core/logging-events-diagnostics/interceptors?WT.mc_id=DT-MVP-5003235); however I'd like to know:

- Is there any equivalent for query filters for UPDATE and DELETE scenarios?
- Has the feature been discussed and rejected/approved as part of the plan?

EF Core version: EF Core 5.0
Database provider: (any - but interested in MS SQL Server and SQLite)
Target framework: .NET 5.0

**Comment by roji** [comment]
@r-aghaei the WHERE clause generated by EF Core in UPDATE and DELETE is generally used only for the key column(s) of the rows to be modified (or to implement optimistic concurrency). The selection of which rows are to be updated/deleted (and therefore any filtering by tenant) is done by the application before this stage, in .NET code, via EF's regular change tracking mechanism.

In other words, with EF Core's "unit of work" pattern, updates/deletes work something like the following:

```c#
// All filtering - including global query filters - get applied when *querying*:
var blogsToBeRemoved = context.Blogs.Where(b => /* some filter */).ToArray();

// Mark all the blogs - which have already been filtered for tenantId - for removal
foreach (var blogToBeRemoved in blogsToBeRemoved)
{
    context.Blogs.Remove(blogToBeRemoved);
}

// Apply the operation to the database. At this point EF Core generates the DELETE SQL with the keys (IDs) of the blogs selected above
context.SaveChanges();
```

To summarize, since with unit-of-work, updating/removal always operates on entities that have already been queried by the application, any filtering happens when querying rather than when saving.

However, when we implement bulk updates (#795), the global filter question indeed becomes relevant there - I've added a note.

**Comment by r-aghaei** [comment]
@roji Thanks for the reply. 

> To summarize, since with unit-of-work, updating/removal always operates on entities that have already been queried by the application, any filtering happens when querying rather than when saving.

That's correct for connected scenarios; as I also mentioned in the question, this is the obvious and common choice, however for a better performance, in disconnected scenarios, it makes sense to have something like query filters.

I'm happy to see it will be considered in bulk operations; however, in general I see it less-related to *bulk operations* but more-related to considerations for *disconnected mode (without loading data)*:

- For bulk operations, you will set the criteria at operation level; quite similar to where clause that you add by `Where` linq method when querying data from DbSets).
- But query filter, are global filters which are applicable at the context level; quite similar to global query filters which are added by `HasQueryFilter`.

The same way that we are able to set global query filters for `SELECT`, I believe it's a good idea to have a feature which allows us to set a global filter for `UPDATE` and `DELETE` statements as well.

**Comment by roji** [comment]
I do see some value in this... however:

* Even in disconnected scenarios, the external entity has to come from somewhere - typically by being queried previously by a different DbContext instance. I'd again expect the query filter to apply at that point, meaning that the application wouldn't typically be attaching entity instances that need to be filtered out. Of course, there's the possibility of a different tenantId (or similar) being configured on the querying and updating context, but that seems more like an issue in the application.
* In other words, can you provide a concrete, convincing scenario where this is needed?
* We're discussing UPDATE and DELETE, but what should be the insertion behavior? Would you argue that if the application adds a new entity the would get excluded by a global query filter, EF Core should ignore that? If not, why not?
* This would be a possibly significant breaking change.

**Comment by r-aghaei** [comment]
**Should EF care about "Global query filter" in INSERT statements?**

I see "Global query filter" as an additional criteria which applies to WHERE clause:
- SELECT, UPDATE, and DELETE statements support WHERE clause, so global query filter makes sense for those.
- INSERT query doesn't support WHERE clause, so global query filter doesn't make sense for it.

**Example  for UPDATE scenario**

A multi-tenant API receives a request to publish a `BlogPost` with `Id=1` and using context information it detects `Tenant1` as the tenant id and setup the context to use a query filter like `TenantId = 'Tenant1'` on `BlogPosts`.

Using the existing features of EF (global query filters only for SELECT), we should first make sure the blog post with id=1 belongs to Tenant1. It means each update scenario, will require an additional SELECT statement followed by the UPDATE statement. Something like this:
```
var entity = db.BlogPosts.Find(id);    // SELECT * FROM BlogPosts WHERE Id=1 AND TenantId = 'Tenant1'
if(entity!=null)                       // If entity found
{
    entity.Published = true;           // Modify the entity
    db.SaveChanges();                  // UPDATE BlogPosts SET Published = 1 WHERE Id = 1
}
```

Using the proposed feature (global query filter for SELECT/UPDATE/DELETE), we don't need the first SELECT statement, because the global query will be applied on UPDATE, which means less communication with Db Server and faster update:
```
var entity = new BlogPosts() { Id = 1} ;
db.Attach(entity);
entity.Published = 1;
db.SaveChanges();                      // UPDATE BlogPosts SET Published = 1 WHERE Id = 1 AND TenantId = 'Tenant1'
```

Using the proposed feature, not only `Publish`  scenario benefits from the feature, all other update and delete scenarios like `Unpublish`, `Edit`, `Delete`, `Archive`, and etc. will benefit from the same central global query filter which has set in DbContext, without requiring an additional DB server communication (the first SELECT statement).

**Comment by roji** [comment]
> I see "Global query filter" as an additional criteria which applies to WHERE clause

I think it's better to look at these questions from a user perspective; so I'd ask myself what the user behavior should be rather, rather than which type of SQL should get affected (implementation detail). For one thing, EF supports non-relational database which don't use SQL at all. 

Regardless, your example for an UPDATE scenario does have merit. One potential issue issue I see, is that we allow users to ignore query filters for particular queries ([see docs](https://docs.microsoft.com/en-us/ef/core/querying/filters#disabling-filters)), but since that depends on IQueryable, it wouldn't work as-is for non-queries.

In any case, we'll discuss this in a team meeting, thanks for taking the time to post!

**Comment by r-aghaei** [comment]
Makes sense. I mean "I as a developer (user of EF)" 😄.

* EF supports non-relational database which don't use SQL at all.* → I think the same argument applies to non-relational DBs as well. I as a developer, most likely expect the filter on RUD, but not C.

Perfect! Thanks for the feedback and for taking time on this 😊.

**Comment by roji** [comment]
@r-aghaei we've discussed this, and think it could have some value as a feature - am putting this in the backlog.

Some notes:
* There are various difficulties with this, e.g. how to distinguish between a concurrent update failure (where the concurrency token causes an update/delete to fail) and the update filter doing its job.
* As this seems to basically be an optimization for a specific disconnected scenario, it's not likely to be highly prioritized.
* This would likely be a separately-defined filter rather than simply applying the existing global query filters, e.g. because doing so would be a breaking change.

**Comment by r-aghaei** [comment]
Great! Makes sense, thanks for the update 😊.

**Comment by r-aghaei** [comment]
@roji Has it been addressed in any of the releases? What's the latest status on this?

**Comment by roji** [comment]
This issue is in the Backlog milestone. This means that it is not planned for the next release (EF Core 8.0). We will [re-assess the backlog](https://docs.microsoft.com/ef/core/what-is-new/release-planning) following the this release and consider this item at that time. However, keep in mind that there are many other high priority features with which it will be competing for resources. Make sure to vote (👍) for this issue if it is important to you.

**Comment by skawtish** [comment]
Any update? Can't believe this isn't in - seems like an oversight. When using multitenant systems its common to partition tables by the tenantid, ignoring this in updates/deletes incurs a larger lookup time if we are required to scan multiple partitions.

**Comment by roji** [comment]
This issue is in the Backlog milestone. This means that it is not planned for the next release (EF Core 8.0). We will [re-assess the backlog](https://docs.microsoft.com/ef/core/what-is-new/release-planning) following the this release and consider this item at that time. However, keep in mind that there are many other high priority features with which it will be competing for resources. Make sure to vote (👍) for this issue if it is important to you.
